File: Array/Sorting/Radix-Sorting/RadixSort.c
#include <stdio.h>

int getMax(int arr[], int n) {
    int max = arr[0];
    for(int i = 1; i < n; i++)
        if(arr[i] > max)
            max = arr[i];
    return max;
}

void countSort(int arr[], int n, int exp) {
    int output[n];
    int count[10] = {0};

    // Count occurrences
    for(int i = 0; i < n; i++)
        count[(arr[i] / exp) % 10]++;

    // Cumulative count
    for(int i = 1; i < 10; i++)
        count[i] += count[i - 1];

    // Build output array
    for(int i = n - 1; i >= 0; i--) {
        int digit = (arr[i] / exp) % 10;
        output[count[digit] - 1] = arr[i];
        count[digit]--;
    }

    // Copy back to original
    for(int i = 0; i < n; i++)
        arr[i] = output[i];
}

int main() {
    int n;
    printf("Enter the size of the array: ");
    scanf("%d", &n);

    int arr[n];
    printf("Enter the elements of the array: ");
    for(int i = 0; i < n; i++)
        scanf("%d", &arr[i]);

    printf("Original array: ");
    for(int i = 0; i < n; i++)
        printf("%d ", arr[i]);

    int max = getMax(arr, n);

    for(int exp = 1; max / exp > 0; exp *= 10)
        countSort(arr, n, exp);

    printf("\nSorted array: ");
    for(int i = 0; i < n; i++)
        printf("%d ", arr[i]);

    return 0;
}

// This code implements the Radix Sort algorithm in C.

File: Array/Sorting/Radix-Sorting/RadixSort.py

# Input array from user
n = int(input("Enter the size of the array: "))
arr = []
print("Enter the elements of the array: ")
for _ in range(n):
    arr.append(int(input()))

print("Original array:", *arr)

# Function to get maximum value
max_val = max(arr)

exp = 1  # Exponent for digit place (1s, 10s, 100s, ...)
while max_val // exp > 0:
    count = [0] * 10
    output = [0] * n

    # Count occurrences of digits at current place
    for num in arr:
        index = (num // exp) % 10
        count[index] += 1

    # Cumulative count
    for i in range(1, 10):
        count[i] += count[i - 1]

    # Build output array
    for i in range(n - 1, -1, -1):
        index = (arr[i] // exp) % 10
        output[count[index] - 1] = arr[i]
        count[index] -= 1

    # Copy output back to arr
    arr = output[:]
    exp *= 10

print("Sorted array: ", *arr)

# This code implements the Radix Sort algorithm in Python.
